File: 20-actors/fuchi/methods/test_vote.py
#!/usr/bin/env python3
"""Tests for 扶持 (fuchi) vote.py — R1(b) 1 SBT = 1 vote + 48h timelock.

Standalone-runnable: python3 test_vote.py
"""
from __future__ import annotations

import sys

from vote import Ballot, ballots_from_seed, cast, finalize, tally


def _ballots(spec, opened=1000):
    out = []
    t = opened + 1
    for did, choice in spec:
        out = cast(out, Ballot(did, choice, t))
        t += 2
    return out


def test_accepts_after_timelock_with_quorum():
    b = _ballots([("did:m:a", "yes"), ("did:m:b", "yes"), ("did:m:c", "yes"), ("did:m:d", "no")])
    r = tally(b, opened_at=1000, now=1060, timelock_h=48)
    assert r["finalizable"] and r["outcome"] == "accepted" and r["yes"] == 3 and r["no"] == 1


def test_pending_before_timelock():
    b = _ballots([("did:m:a", "yes"), ("did:m:b", "yes"), ("did:m:c", "yes")])
    r = tally(b, opened_at=1000, now=1010, timelock_h=48)
    assert r["finalizable"] is False and r["outcome"] == "pending"


def test_finalize_raises_before_timelock():
    b = _ballots([("did:m:a", "yes"), ("did:m:b", "yes")])
    try:
        finalize(b, opened_at=1000, now=1010, timelock_h=48)
    except ValueError as e:
        assert "timelock" in str(e)
        return
    raise AssertionError("finalize must raise before the timelock elapses")


def test_rejected_without_quorum():
    b = _ballots([("did:m:a", "yes")])
    r = tally(b, opened_at=1000, now=1060, timelock_h=48, quorum=3)
    assert r["quorum_met"] is False and r["outcome"] == "rejected"


def test_rejected_when_no_beats_yes():
    b = _ballots([("did:m:a", "yes"), ("did:m:b", "no"), ("did:m:c", "no")])
    r = tally(b, opened_at=1000, now=1060, timelock_h=48)
    assert r["outcome"] == "rejected"


def test_one_sbt_one_vote_rejects_duplicate():
    b = cast([], Ballot("did:m:a", "yes", 1001))
    try:
        cast(b, Ballot("did:m:a", "no", 1002))
    except ValueError as e:
        assert "1 SBT = 1 vote" in str(e)
        return
    raise AssertionError("a duplicate voter must be rejected (1 SBT = 1 vote)")


def test_ballot_weight_must_be_one():
    try:
        Ballot("did:m:a", "yes", 1001, weight=5)
    except ValueError as e:
        assert "1 SBT = 1 vote" in str(e)
        return
    raise AssertionError("weight != 1 must be rejected")


def test_server_voter_unrepresentable():
    try:
        Ballot("server", "yes", 1001)
    except ValueError as e:
        assert "G9" in str(e) or "G4" in str(e)
        return
    raise AssertionError("a server voter must be rejected (no-server-key)")


def test_ballot_server_key_refused():
    try:
        Ballot("did:m:a", "yes", 1001, server_held_key=True)
    except ValueError as e:
        assert "no-server-key" in str(e)
        return
    raise AssertionError("server-held-key ballot must be refused (G9)")


def test_out_of_window_ballot_not_counted():
    # a ballot cast after the window closes is ignored in the tally
    b = [Ballot("did:m:a", "yes", 1004), Ballot("did:m:b", "yes", 1004),
         Ballot("did:m:c", "yes", 1004), Ballot("did:m:d", "yes", 9999)]
    r = tally(b, opened_at=1000, now=1060, timelock_h=48)
    assert r["yes"] == 3   # the 9999 ballot is outside [1000, 1048]


def test_ballots_from_seed_dedupes():
    recs = [{":ballot/voter": "did:m:a", ":ballot/choice": ":yes", ":ballot/cast-at": 1004},
            {":ballot/voter": "did:m:b", ":ballot/choice": ":no", ":ballot/cast-at": 1006}]
    b = ballots_from_seed(recs)
    assert len(b) == 2 and b[0].choice == "yes"


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"test_vote.py: {len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/fuchi/methods/test_vote.clj
;; test_vote.clj — 扶持 vote: 1 SBT=1 vote + 48h timelock + quorum, parity with vote.py;
;; + live-gate R2. Run via `bb test:fuchi`. ADR-2606142300.
(ns fuchi.methods.test-vote
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [fuchi.methods.vote :as v]
            [fuchi.methods.live-gate :as lg]))

(defn- mk [voter choice at] (v/make-ballot {:voter-did voter :choice choice :cast-at at}))

(def ^:private ballots
  (-> [] (v/cast (mk "v1" "yes" 110)) (v/cast (mk "v2" "yes" 120))
         (v/cast (mk "v3" "no" 130))  (v/cast (mk "v5" "abstain" 140))
         (v/cast (mk "v4" "yes" 200))))                    ; v4 outside the [100,148] window

(deftest tally-window-quorum-outcome
  (testing "only in-window ballots count; quorum + yes>no → accepted (golden from vote.py)"
    (let [t (v/tally ballots 100 150)]                     ; close=148, now=150 finalizable
      (is (= 2 (:yes t)))
      (is (= 1 (:no t)))
      (is (= 1 (:abstain t)))
      (is (= 4 (:voters t)))                               ; v4@200 excluded
      (is (= 148 (:close t)))
      (is (true? (:quorum-met t)))
      (is (true? (:finalizable t)))
      (is (= "accepted" (:outcome t))))))

(deftest pending-before-timelock
  (testing "before the 48h window closes the outcome is pending (never early)"
    (is (= "pending" (:outcome (v/tally ballots 100 120))))
    (is (false? (:finalizable (v/tally ballots 100 120))))))

(deftest thin-vote-rejected
  (testing "quorum not met → rejected, never auto-accepted"
    (let [thin (-> [] (v/cast (mk "a" "yes" 110)) (v/cast (mk "b" "yes" 120)))]
      (is (= "rejected" (:outcome (v/tally thin 100 150)))))))   ; 2 < quorum 3

(deftest one-sbt-one-vote
  (testing "a duplicate voter DID is rejected at cast time (1 SBT = 1 vote)"
    (is (thrown? Exception (v/cast ballots (mk "v1" "no" 125))))))

(deftest ballot-invariants
  (testing "weight 1, no-server-key, no :server/:anon voter, valid choice"
    (is (thrown? Exception (v/make-ballot {:voter-did "z" :choice "yes" :cast-at 1 :weight 2})))
    (is (thrown? Exception (v/make-ballot {:voter-did "z" :choice "yes" :cast-at 1 :server-held-key true})))
    (is (thrown? Exception (v/make-ballot {:voter-did "server" :choice "yes" :cast-at 1})))
    (is (thrown? Exception (v/make-ballot {:voter-did "did:server:x" :choice "yes" :cast-at 1})))
    (is (thrown? Exception (v/make-ballot {:voter-did "anon" :choice "yes" :cast-at 1})))
    (is (thrown? Exception (v/make-ballot {:voter-did "z" :choice "maybe" :cast-at 1})))
    (is (= "yes" (:choice (v/make-ballot {:voter-did "z" :choice ":yes" :cast-at 1}))))))   ; colon stripped

(deftest finalize-strict-timelock
  (testing "finalize RAISES before the timelock; succeeds after"
    (is (thrown? Exception (v/finalize ballots 100 120)))        ; 120 < 148
    (is (= "accepted" (:outcome (v/finalize ballots 100 150))))))

(deftest finalize-binding-r2
  (testing "binding finalize via the R2 autonomous gate; timelock still strict"
    (let [g  (lg/make-gate {:leg "vote"})
          fb (v/finalize-binding ballots 100 150 g)]
      (is (= "accepted" (:outcome fb)))
      (is (true? (:binding fb)))
      (is (= 7 (:council-level fb)))
      (is (thrown? Exception (v/finalize-binding ballots 100 120 g))))))  ; gate can't bypass timelock

(deftest live-gate-r2
  (testing "live-gate: known legs admissible (R2 autonomous); unknown leg raises"
    (is (true? (:admissible (lg/gate-status (lg/make-gate {:leg "provision"})))))
    (is (= 7 (second (lg/leg-policy "couple"))))             ; couple needs Lv7
    (is (thrown? Exception (lg/make-gate {:leg "bribe"})))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'fuchi.methods.test-vote)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
